libparsec/crates/client/src/invite/greeter.rs

```diff
@@ -253,6 +262,135 @@ impl From<ConnectionError> for GreetInProgressError {
     }
 }
 
+// Cancel greeting attempt helper
+
+async fn cancel_greeting_attempt(
+    cmds: &AuthenticatedCmds,
+    greeting_attempt: GreetingAttemptID,
+    reason: CancelledGreetingAttemptReason,
+) -> Result<(), GreetInProgressError> {
+    use authenticated_cmds::latest::invite_greeter_cancel_greeting_attempt::{Rep, Req};
+
+    let req = Req {
+        greeting_attempt,
+        reason,
+    };
+    let rep = cmds.send(req).await?;
+
+    match rep {
+        Rep::Ok => Ok(()),
+        // Expected errors
+        Rep::InvitationCompleted => Err(GreetInProgressError::AlreadyDeleted),
+        Rep::InvitationCancelled => Err(GreetInProgressError::AlreadyDeleted),
+        Rep::AuthorNotAllowed => Err(GreetInProgressError::GreeterNotAllowed),
+        Rep::GreetingAttemptAlreadyCancelled {
+            origin,
+            reason,
+            timestamp,
+        } => Err(GreetInProgressError::GreetingAttemptCancelled {
+            origin,
+            reason,
+            timestamp,
+        }),
+        // Unexpected errors
+        Rep::GreetingAttemptNotFound => Err(anyhow::anyhow!("Greeting attempt not found").into()),
+        Rep::GreetingAttemptNotJoined => Err(anyhow::anyhow!("Greeting attempt not joined").into()),
+        rep @ Rep::UnknownStatus { .. } => {
+            Err(anyhow::anyhow!("Unexpected server response: {:?}", rep).into())
+        }
+    }
+}
+
+// Greeter step helper
+
+static STEP_THROTTLE: Duration = Duration::milliseconds(100);
+
+async fn run_greeter_step_until_ready(
+    cmds: &AuthenticatedCmds,
+    greeting_attempt: GreetingAttemptID,
+    greeter_step: invite_greeter_step::GreeterStep,
+    time_provider: &TimeProvider,
+) -> Result<invite_greeter_step::ClaimerStep, GreetInProgressError> {
+    let mut last_call = Option::None;
+    let req = invite_greeter_step::Req {
+        greeting_attempt,
+        greeter_step,
+    };
+
+    // Loop over the requests
+    loop {
+        // Throttle the requests
+        if let Some(last_call) = last_call {
+            let duration = last_call + STEP_THROTTLE - time_provider.now();
+            time_provider.sleep(duration).await;
+        }
+        last_call = Some(time_provider.now());
```

Comment:
Shouldn't this kind of code (or at least the `STEP_THROTTLE`) be factorized with identical code in `invite/claimer.rs`? Or we really want to keep the flexibility of independent step/algorithm?

Reply:
I'll look into it.

Reply:
5d2b34fdfa2419fc899db0b08725632b62b1832b